### DISCUSSION TITLE
Possibility of migrating the back all supported in api/rest with nodejs?

### DISCUSSION THREAD

**Discussion by WelingtonMonteiro** [RetroAchievements/RAWeb#865] [Ideas]
Thus, we totally decouple technologies, enabling the construction of mobile applications, using the new api.
We could migrate the backend to nodejs and api first, and keep the ui in PHP.
I'm wanting to create a hybrid mobile app for the site, but without an api it would be limited, so moving the backend to full api would be a good thing, with that I could also upgrade the UI to a 2.0 version with other technologies if applicable and or framework.

**Comment by meleu** [comment]
Although I'm not currently with enough freetime, I would be very much interested in creating/working/contributing with a NodeJS API.

**Comment by luchaos** [comment]
We had that discussion in the past iirc. While i can see us replacing some services with other runtimes in the future - we really should focus on separating the services of the current site first.

The Laravel framework iteration will provide just that - separating it into site, connect api and web api.
This feature seems to revolve solely around the web api part which is easily implemented in Laravel as it keeps the models and routing together and reusable for the other services.

Running it via Laravel Octane will provide an noteworthy performance boost, too.

It's far too soon to go the node, or even other runtime route, imo. Not sure Node is the best fit either - the dependency management is still a mess, riddled with supply chain attacks. I'd personally rather go with deno, if even.

Providing a restful web api is just one of many features that have to be covered by the choices of the runtime.

As reference https://github.com/RetroAchievements/RAWeb/discussions/652

**Reply by meleu** [reply]
I must confess that the interest I have shown above is NOT guided by a weighted technical assessment. But out of a (maybe egoistical) interest in taking advantage of the knowledge of the technologies I've been working with recently.

Thankfully we have experienced people with more mileage on board. :pray: 

By the way, using Deno could be fun also. Maybe that would label us as "early adopters". :sophisticated:

**Comment by WelingtonMonteiro** [comment]
I mentioned nodejs above because it's the framework I've mastered the most, I've been working with it for over 7 years, so I have solid knowledge, besides being javascript so other devs could contribute or try to contribute because it makes the learning curve easier, I'm currently reading and trying to understand the web project currently to try to separate requests in api.

At first I want to create all site requests in apis, and use an ORM to connect to the mysql database, thus decoupling all the dependency on SQL code written in the code, raising the level at this point, we can use the sequelize that does this very well, alias I did a test here after configuring the database in the docker, and with a command already extracts all the tables for sequelize.

As for Deno, I've worked with it a little while ago, and I wouldn't go from Deno yet, as the framework is very new, that's my opinion.
![image](https://user-images.githubusercontent.com/4640319/144837099-31a0d6dc-44ce-49f2-b1c2-15563e155359.png)

At first I separated into modules, so everything related to the module would be in its respective domain/folder.

The idea would be to work on each module and migrate the apis and SQL commands to sequelize, ending up it would go up and change the site's current request to this ready-made api.

and with that, gradually migrating each call from the front to the new apis and disabling the old ones.

**Reply by WelingtonMonteiro** [reply]
I saw in other discussions that they were working since 2018 on a 2.0 version of the project, I saw it in your MELEU on Thoreau's podcast, talking about this 2.0 version but I didn't find code or anything about the current progress of the project, if they are still working on it, if they stopped, etc.

**Reply by luchaos** [reply]
I'm still working on it, yes - well, a team by now, actually.

I think it's great that you put work into it but one thing to consider is that the v2 project does a lot more than to just rewrite what's currently there. It completely changes the underlying data structures as those are the root cause for many problems  the community has had over the years. The rewrite does not happen just for the sake of changing the stack but is a necessity. At this point I would not advise to introduce another stack but finish the structural changes first. This will take some more time.

**Reply by WelingtonMonteiro** [reply]
I understand, I know v2 is not public, and I know the current priority is to change the structure on the front, but it would be interesting to start migrating calls to api, or less centralize requests in a folder/module to facilitate later migration of this part, i don't master PHP or laravel, so i can't give my opinion at this point, but i would like to be helping in something, mainly in this part of api migration.
